### PR TITLE
[C++17] Support __GCC_[CON|DE]STRUCTIVE_SIZE

### LINKED ISSUE #60174: FR: Add cacheline size macros sufficient to support C++17 std::hardware_{con,de}structive_interference_size
libc++ implements std::hardware_{con,de}structive_interference_size on gcc only due to clang lacking sufficient macro support. This bug requests that clang add sufficient support to implement this C++17ism in libc++.

### PR DESCRIPTION
**Pull request by AaronBallman**
These macros are used by STL implementations to support implementation of std::hardware_destructive_interference_size and std::hardware_constructive_interference_size

Fixes #60174

### CODE REVIEW

**Comment by AaronBallman** [comment]
During today's office hours, @philnik777 mentioned that support for these macros would be very beneficial to libc++ folks. This is intended as a starting point for the support in that it does not currently have any special logic for any particular target and it does not have a diagnostic for using these macros within a header file.

In support of getting some target-specific values that make sense, I'm hopeful that the following folks can suggest proper values for the targets they know about (I pulled this list from the LLVM code owners list, so if there are others I should CC, please add them): @arsenm @asb @asl @jholewinski @mshockwave @SixWeining @benshi001 @kaz7 @TNorthover @nigelp-xmos @jpienaar @phoebewang @andreisfr @markschimmel @uweigand @wuzish @iliya-diyachkov 

If any of you are able to let me know what the minimum offset is between two objects to avoid false sharing and what the maximum size of contiguous memory is to promote true sharing for your target, I can add it as part of this PR. If you don't know or are happy with the defaults I picked, that's fine too. I figure we can improve these values as we see fit. And if the answer is "it's complicated", I'd appreciate if we could handle that in a follow-up patch so we can get libc++ folks unblocked.

**Comment by phoebewang** [comment]
Sorry, I know nothing about it. But it looks to me it's to match with GCC, why don't borrow the value from GCC as a beginning?

**Comment by cor3ntin** [comment]
@AaronBallman Did you read https://discourse.llvm.org/t/rfc-c-17-hardware-constructive-destructive-interference-size/48674/42 ? I want to make sure we do abide by it and it's unclear a consensus was ever formed or called

**Comment by uweigand** [comment]
For SystemZ the correct value is 256.   In general I agree it makes sense to look at the GCC implementation as a source of reasonable values.   Also, I think there probably should be no generic default value at all - it there is no platform-specific value known, it seems better to not define those values rather than define them to some incorrect value ...

**Comment by arsenm** [comment]
For AMDGPU 64 is probably right

**Comment by AaronBallman** [comment]
> @AaronBallman Did you read https://discourse.llvm.org/t/rfc-c-17-hardware-constructive-destructive-interference-size/48674/42 ? I want to make sure we do abide by it and it's unclear a consensus was ever formed or called

Yes, I looked at the thread and while there are very valid criticisms of what was standardized, the end result is still that we need this support 1) for Clang to support libc++ implementing a C++17 feature for conformance reasons and 2) for GCC compatibility (re: the macros that are exposed). We can either let libc++ solve this problem entirely on their own and not aim for GCC compatibility here, or we can bite the bullet and do our best. Given that GCC exposes these macros, I think it makes more sense for Clang to predefine them than for libc++ to try to do this from their end.

**Comment by AaronBallman** [comment]
> Sorry, I know nothing about it. But it looks to me it's to match with GCC, why don't borrow the value from GCC as a beginning?

Because the values depend on the target and the compiler options picked. I don't have that information myself, and given how many targets and options we support, I am hoping target owners can help rather than having to guess.

**Comment by AaronBallman** [comment]
> For SystemZ the correct value is 256.

Thanks! Double-checking: for both constructive and destructive?

> In general I agree it makes sense to look at the GCC implementation as a source of reasonable values. Also, I think there probably should be no generic default value at all - it there is no platform-specific value known, it seems better to not define those values rather than define them to some incorrect value ...

On the one hand, I agree. But then libc++ will still have to pick default values to expose (these are `constexpr size_t` variables in libc++ rather than macros or `std::optional` values), so that just moves the problem elsewhere. Also, as best I can tell, that doesn't seem to be how GCC behaves (at least, from trying various GCC flavors on Compiler Explorer).

**Comment by AaronBallman** [comment]
> > > For SystemZ the correct value is 256.
> > 
> > 
> > Thanks! Double-checking: for both constructive and destructive?
> 
> Yes, for both. Every SystemZ model (supported by GCC/LLVM) has a L1 cache line size of 256 bytes.

Excellent, thank you!

> > > In general I agree it makes sense to look at the GCC implementation as a source of reasonable values. Also, I think there probably should be no generic default value at all - it there is no platform-specific value known, it seems better to not define those values rather than define them to some incorrect value ...
> > 
> > 
> > On the one hand, I agree. But then libc++ will still have to pick default values to expose (these are `constexpr size_t` variables in libc++ rather than macros or `std::optional` values), so that just moves the problem elsewhere. Also, as best I can tell, that doesn't seem to be how GCC behaves (at least, from trying various GCC flavors on Compiler Explorer).
> 
> Looks like GCC's libstdc++ will simply not provide those variables either if GCC does not know those values for the current target:
> 
> ```
> #ifdef __cpp_lib_hardware_interference_size // C++ >= 17 && defined(gcc_dest_sz)
>   inline constexpr size_t hardware_destructive_interference_size = __GCC_DESTRUCTIVE_SIZE;
>   inline constexpr size_t hardware_constructive_interference_size = __GCC_CONSTRUCTIVE_SIZE;
> #endif // __cpp_lib_hardware_interference_size
> ```

@ldionne do you have a preference? I can always skip defining the predefined macros unless a target has opted into providing that information.

**Comment by cor3ntin** [comment]
I think it makes sense to go with this patch and let target maintainers tweak the value subsequently.
It seems that trying to not provide a value unless explicitly specified by the target would be worse for library vendors as
`hardware_destructive_interference_size` is specified to exist unconditionally.

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:99]: > These macros can be  overridden on the command line with ``-D``, if desired. 

I would rather we don't say that at it could lead to ODR violations.
IE consider `alignas(std::hardware_destructive_interference_size()) struct S{};` with different values in different TUs

Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst:99]: Hmm, given that I'm not implementing the on-by-default diagnostic at this time, it's probably reasonable to remove that bit. Thanks!

**Comment by ldionne** [comment]
> > > > For SystemZ the correct value is 256.
> > > 
> > > 
> > > Thanks! Double-checking: for both constructive and destructive?
> > 
> > 
> > Yes, for both. Every SystemZ model (supported by GCC/LLVM) has a L1 cache line size of 256 bytes.
> 
> Excellent, thank you!
> 
> > > > In general I agree it makes sense to look at the GCC implementation as a source of reasonable values. Also, I think there probably should be no generic default value at all - it there is no platform-specific value known, it seems better to not define those values rather than define them to some incorrect value ...
> > > 
> > > 
> > > On the one hand, I agree. But then libc++ will still have to pick default values to expose (these are `constexpr size_t` variables in libc++ rather than macros or `std::optional` values), so that just moves the problem elsewhere. Also, as best I can tell, that doesn't seem to be how GCC behaves (at least, from trying various GCC flavors on Compiler Explorer).
> > 
> > 
> > Looks like GCC's libstdc++ will simply not provide those variables either if GCC does not know those values for the current target:
> > ```
> > #ifdef __cpp_lib_hardware_interference_size // C++ >= 17 && defined(gcc_dest_sz)
> >   inline constexpr size_t hardware_destructive_interference_size = __GCC_DESTRUCTIVE_SIZE;
> >   inline constexpr size_t hardware_constructive_interference_size = __GCC_CONSTRUCTIVE_SIZE;
> > #endif // __cpp_lib_hardware_interference_size
> > ```
> 
> @ldionne do you have a preference? I can always skip defining the predefined macros unless a target has opted into providing that information.

Not defining these variables when we don't have a good value for them would make us non-conforming, so I agree with @cor3ntin : I think we should define these macros all the time and simply document that it's a best effort thing. This whole feature is a best effort thing anyways.


> I think it makes sense to go with this patch and let target maintainers tweak the value subsequently. It seems that trying to not provide a value unless explicitly specified by the target would be worse for library vendors as `hardware_destructive_interference_size` is specified to exist unconditionally.

Exactly. It's arguable whether this was a good decision on WG21's end or not, but it was specified as provided unconditionally by a conforming implementation, therefore that's what we need to do (IMO).

File: clang/docs/ReleaseNotes.rst
Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst]: Done

**Review by cor3ntin — APPROVED** [review]
We probably want more feedback on that but otherwise, LGTM

**Comment by AaronBallman** [comment]
Here are the values that I found for GCC that don't match the defaults.

|Platform|Constructive|Destructive|Link
|-|-|-|-|
|ARM64|64|256|https://godbolt.org/z/G7arf9nPP|
|AVR|32|32|https://godbolt.org/z/Yq64f9Koj|
|BPF|32|32|https://godbolt.org/z/zezK71zoe|
|M68K|32|32|https://godbolt.org/z/h9f4vaYcs|
|mips|32|32|https://godbolt.org/z/v3nfPMx9b|
|mips64|32|32|https://godbolt.org/z/7G66rc4nY|
|power|32|32|https://godbolt.org/z/bWjdosE3f|
|power64|128|128|https://godbolt.org/z/jT86TMYTK|
|RISC-V 32-bit|32|32|https://godbolt.org/z/neErxEbWP|
|RISV-V 64-bit|32|32|https://godbolt.org/z/nG5zjjxdo|
|s390x|256|256|https://godbolt.org/z/j34f3eEMz|
|sparc|32|32|https://godbolt.org/z/96hscbesM|

Note, this is testing default configuration with no CPU tuning. I'll nab the values from here that I can map to our targets but target owners should validate these values and let me know if anything looks amiss.

**Comment by philnik777** [comment]
@AaronBallman thanks for picking this up! I agree with @cor3ntin and @ldionne that we should provide the macros unconditionally. This brings us one step closer to finally making libc++ fully conforming in C++17. With the PSTL finally making steady progress, we're probably only missing a few audits and the math special functions to be fully conforming.

**Comment by AaronBallman** [comment]
Precommit CI is currently failing on libc++ tests that are impacted by this change; should I solve those as part of this patch or are you okay with fixing those after this patch lands? (Or would you like to push a fix to this branch?)

**Review by ldionne — APPROVED** [review]
I'd rather keep the libc++ tests green while landing this PR. I tried something out, let's see if that works.

Otherwise LGTM, thanks for tackling this long-standing item on our TODO list :-)

File: clang/docs/ReleaseNotes.rst
Comment (jyknight) [reply clang/docs/ReleaseNotes.rst]: I disagree with this resolution.

In order for these values to be at all useful, we MUST reserve the ability to change these values arbitrarily based on CPU tuning flags or in new compiler versions. If we end up having a release (of clang OR libc++) that exposes these values without a warning, I think that is indeed likely to result in users triggering ODR problems. 

But, the answer to this problem must be to implement a diagnostic to try to avoid ODR issues, not to hide that the values might vary across TUs in the documentation.

Comment (cor3ntin) [reply clang/docs/ReleaseNotes.rst]: My comment was only to not encourage using `-D`. The fact it's defined through a macro is really an implementation details, the macro is not meant to be user facing.

As for whether we need something specific for ODR here... we certainly do not for other abi-impacting feature.
you can already get into issues if the target is different in different TUs. Or if any define is different across TUs
(C++ modules already catch that sort of ODR issues when use in alignas etc)

I'm also not sure whether we usually warn when undocumented macros are used.

**Review by jyknight — COMMENTED** [review]
I  think having the on-by-default diagnostic before we release this functionality is critically important -- but the primarily-useful part of that is going to be for the public libc++ APIs which expose these values (and, that already landed in mainline). This PR is almost an internal implementation detail, so doesn't really make things much worse than they are, so I don't feel like it needs to be held up by that.

I do really hope that diagnostic does get added soon.

I concur with previous comments that getting the values exactly right for each target in the first commit isn't super-important: each target maintainer may adjust as desired in the future, since these are abi-unstable values.

However, I do note that we already _almost_ have the required data in the LLVM backend for many targets, via TargetTransformInfo::getCacheLineSize. That is only a single value, so it's only correct if requesting tuning for a single CPU-model, vs a generic tuning which must express the full range of possibilities. But, when collecting the values to use here, that existing data has been ignored, which I think is unfortunate. Ideally, we would actually _share_ the data, rather than duplicating it in two different places. Though, due to architectural decisions made way-back-in-time, we often have trouble achieving actual sharing of backend-specific data...

But at the least, if we cannot actually share the same code, maybe we should attempt to be internally-consistent?

File: clang/docs/ReleaseNotes.rst
Comment (jyknight) [reply clang/docs/ReleaseNotes.rst]: > My comment was only to not encourage using -D. The fact it's defined through a macro is really an implementation details, the macro is not meant to be user facing.

There should to be a user-facing way to set this, in order to pin it to a particular value. If not -D, then e.g. copy GCC's `--param hardware_destructive_interference_size=N` command-line.

> we don't really make special cases for all the other potentially ABI-terrible ideas people can come up with in a header file

IMO, the special thing about this is that it's exposed by the standard library in a way that invites and suggests the _precise usage_ which is ABI-unstable. The purpose of `hardware_destructive_interference_size` is to put `alignas(std::hardware_destructive_interference_size)` on a struct. Yet, putting that into a header is a terrible idea almost all of the time -- it should be affected by -mtune, and may be changed by compiler upgrades. This is quite non-obvious issue for users, since the C++ stdlib generally goes to a huge amount of effort to remain ABI-stable for users. How do they know that _this_ constant is not ABI-stable, unlike the entire rest of the library?

The way GCC handles it is to warn for _any_ use of `std::hardware_destructive_interference_size` in a header -- but only that one, not `hardware_constructive_interference_size`. I think the rationale is that the documented/expected usage of `destructive` is to use to add padding and/or alignment, while the documented/expected usage of `constructive` is to static_assert that a struct is smaller than it. It actually doesn't 
bother to warn on the internal __GCC_* macros at all.

That may not be ideal, but it's a relatively simple solution that seems likely to be good enough. I think Clang could quite reasonably do the same thing.

Comment (AaronBallman) [reply clang/docs/ReleaseNotes.rst]: > There should to be a user-facing way to set this, in order to pin it to a particular value. If not -D, then e.g. copy GCC's `--param hardware_destructive_interference_size=N` command-line.

`-D` is my preference, mostly because `--param` is not something we already support for options. I could put something into the extensions documentation to mention this?

> How do they know that this constant is not ABI-stable, unlike the entire rest of the library?

I would assume that anyone using this interface is aware of the fact that it's tightly coupled with the CPU tuning used by the program and I would hope that folks are compiling with sanitizers enabled so they'd catch ABI issues more broadly than just this one narrow case. At the same time, defense in depth is not a bad thing.

> That may not be ideal, but it's a relatively simple solution that seems likely to be good enough. I think Clang could quite reasonably do the same thing.

I'll take a run at implementing that; it feels a bit user hostile because there are benign uses within a header file and the macro is exposed in C while the STL interface isn't, but it does provide *some* measure of protection.

File: clang/include/clang/Basic/TargetInfo.h
Comment (AaronBallman) [reply clang/include/clang/Basic/TargetInfo.h:1795]: Good call! Done.

**Comment by cor3ntin** [comment]
For the record, I am fairly opposed to a warning on normal use of a standard function.

I'm not very concerned about users using different target flags across TUs because this is already a massive foot gun and I do not buy that we are making things materially worse.

Instead, by having opinionated on-by-default warnings we encourage users to not care or proactively distrust our warnings. Users will use that functions in headers as a lot of people like to put most of their implementation in headers.

Don't get me wrong, I think diagnosing actual ODR-violation is a worthwhile goal, and as @AaronBallman mentioned
there are ways to do that (modules, asan, build system warning on inconsistent flags), but from within a single TU
we don't have a way to detect actual ODR issues, so we would end up forcing our opinions on users, which I'm not sure is marginally better than not implementing the feature at all.

I don't think this feature warrants deviating from our policy that on-by default warnings should be actionable and have a low false-positive rate, nor the complexity to actually produce the warning.

If we get users reporting issue with that feature, then we can always emit a deprecation warning.

**Comment by AaronBallman** [comment]
> I'd rather keep the libc++ tests green while landing this PR. I tried something out, let's see if that works.

It looks like some further adjustments may be needed as some stage1 builders are failing:
```
# .---command stderr------------
# | /home/runner/_work/llvm-project/llvm-project/libcxx/test/std/language.support/support.dynamic/hardware_inference_size.compile.pass.cpp:16:32: error: no member named 'hardware_destructive_interference_size' in namespace 'std'
# |    16 | ASSERT_SAME_TYPE(decltype(std::hardware_destructive_interference_size), const std::size_t);
# |       |                           ~~~~~^
# | /home/runner/_work/llvm-project/llvm-project/libcxx/test/support/test_macros.h:271:48: note: expanded from macro 'ASSERT_SAME_TYPE'
# |   271 |     static_assert((test_macros_detail::is_same<__VA_ARGS__>::value), \
# |       |                                                ^~~~~~~~~~~
# | /home/runner/_work/llvm-project/llvm-project/libcxx/test/std/language.support/support.dynamic/hardware_inference_size.compile.pass.cpp:17:32: error: no member named 'hardware_constructive_interference_size' in namespace 'std'
# |    17 | ASSERT_SAME_TYPE(decltype(std::hardware_constructive_interference_size), const std::size_t);
# |       |                           ~~~~~^
# | /home/runner/_work/llvm-project/llvm-project/libcxx/test/support/test_macros.h:271:48: note: expanded from macro 'ASSERT_SAME_TYPE'
# |   271 |     static_assert((test_macros_detail::is_same<__VA_ARGS__>::value), \
# |       |                                                ^~~~~~~~~~~
# | 2 errors generated.
# `-----------------------------
```
I'm not quite certain what's causing the failure though

> Otherwise LGTM, thanks for tackling this long-standing item on our TODO list :-)

Happy to help!

**Comment by ldionne** [comment]
> > I'd rather keep the libc++ tests green while landing this PR. I tried something out, let's see if that works.
> 
> It looks like some further adjustments may be needed as some stage1 builders are failing:
> 
> ```
> [...]
> ```

Ah, yes, we're using the nightly clang which is labeled as clang-19 but doesn't contain the change yet. I'll mark the test as `UNSUPPORTED` on all Clang flavors for now (which is how the test should have been marked), and we will fix that in the patch that implements the library feature.

**Comment by AaronBallman** [comment]
I think this resolves all the outstanding issues. I plan to land these changes sometime today unless I hear otherwise (we can also address additional concerns post-commit).

File: clang/docs/LanguageExtensions.rst
Comment (cor3ntin) [inline clang/docs/LanguageExtensions.rst:5589]:
```suggestion
**Note: the values the macros expand to are not guaranteed to be stable. They are are affected by architectures and CPU tuning flags, can change between releases of Clang and will not match the values
defined by other compilers such as GCC.
Compiling different TUs depending on these flags (or `std::hardware_constructive_interference` ,  `std::hardware_destructive_interference`)  with different compilers, defines or architecture flags will lead to ODR violations bugs and should be avoided.
```

Comment (AaronBallman) [reply clang/docs/LanguageExtensions.rst]: Done!

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by kxxt** [comment]
> Here are the values that I found for GCC that don't match the defaults.
> Platform 	Constructive 	Destructive 	Link
> ARM64 	64 	256 	https://godbolt.org/z/G7arf9nPP
> AVR 	32 	32 	https://godbolt.org/z/Yq64f9Koj
> BPF 	32 	32 	https://godbolt.org/z/zezK71zoe
> M68K 	32 	32 	https://godbolt.org/z/h9f4vaYcs
> mips 	32 	32 	https://godbolt.org/z/v3nfPMx9b
> mips64 	32 	32 	https://godbolt.org/z/7G66rc4nY
> power 	32 	32 	https://godbolt.org/z/bWjdosE3f
> power64 	128 	128 	https://godbolt.org/z/jT86TMYTK
> RISC-V 32-bit 	32 	32 	https://godbolt.org/z/neErxEbWP
> RISV-V 64-bit 	32 	32 	https://godbolt.org/z/nG5zjjxdo
> s390x 	256 	256 	https://godbolt.org/z/j34f3eEMz
> sparc 	32 	32 	https://godbolt.org/z/96hscbesM
> 
> Note, this is testing default configuration with no CPU tuning. I'll nab the values from here that I can map to our targets but target owners should validate these values and let me know if anything looks amiss.

FYI [the value 32 is the default value set by GCC and might be wrong](https://gcc.gnu.org/pipermail/libstdc++/2024-September/059490.html). I have submitted [a bug to gcc about the wrong cacheline size 32 for riscv64 target](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=116662).

**Pull request merged**
